category_encoders/target_encoder.py

```diff
@@ -4,6 +4,7 @@
 from category_encoders.ordinal import OrdinalEncoder
 import category_encoders.utils as util
 import warnings
+import flatten_dict as fd
```

Comment:
this is also the reason why the tests are failing at the moment

Reply:
I think the flatten-dict function is complete and possibly too extensive to re-write.  We could ask the author if we could copy the required function into our utils, but I'd hesitate to re-write it.  We tried and found flatten-dict to be better.  Will add it into setiup.py for the moment, until we reach concensus.

Reply:
I think I just found a neat way:  
pandas apparently implements a function `json_normalize` that can be used to flatten dicts as well. This is very well maintained. For reference it's way number 3 here: https://www.freecodecamp.org/news/how-to-flatten-a-dictionary-in-python-in-4-different-ways/
I just tried it out and it seems to work:  

With this function
```python
def flatten_reverse_dict(d):
    sep = "___"
    [flat_dict] = pd.json_normalize(d, sep=sep).to_dict(orient='records')
    reversed_flat_dict = {v: tuple(k.split(sep)) for k,v in flat_dict.items()}
    return reversed_flat_dict
```  
we get the same results as otherwise  
```python
hierarchy_multi_level = {
    'Animal': {
        'Warm-Blooded':
            {'Mammals': ('Cat', 'Dog'),
             'Birds': ('Osprey', 'Kite'),
             'Fish': ('Carp', 'Clownfish')
             },
        'Cold-Blooded':
            {'Reptiles': ('Lizard'),
             'Amphibians': ('Snake', 'Frog')
             }
    }}

flattend = flatten(hierarchy_multi_level["Animal"], inverse=True)
flattend_pd = flatten_reverse_dict(hierarchy_multi_level["Animal"])
print(flattend_pd == flattend)  # prints True
``` 
This would be just 4 lines.  
What do you think? Is this an acceptable solution for you?

Reply:
A very nice solution!  I have tested it, so will push it to the branch now.  Nice work!